Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -0,0 +1,469 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+using QuantConnect.Configuration;
+using QuantConnect.Data.Market;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Orders;
+using QuantConnect.Securities;
+using QuantConnect.Util;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using TradingApi.Bitfinex;
+using TradingApi.ModelObjects.Bitfinex.Json;
+
+namespace QuantConnect.Brokerages.Bitfinex
+{
+
+    /// <summary>
+    /// Bitfinex exchange REST integration.
+    /// </summary>
+    public partial class BitfinexBrokerage : Brokerage, IDataQueueHandler
+    {
+
+        #region Declarations
+        /// <summary>
+        /// Ticks collection
+        /// </summary>
+        protected List<Tick> ticks = new List<Tick>();
+        CancellationTokenSource _tickerToken;
+        /// <summary>
+        /// Divisor for prices. Scales prices/volumes to allow trades on 0.01 of unit
+        /// </summary>
+        protected decimal divisor = 100;
+        readonly object _fillLock = new object();
+        const string buy = "buy";
+        const string sell = "sell";
+        /// <summary>
+        /// Currently limited to BTCUSD
+        /// </summary>
+        protected Symbol symbol = Symbol.Create("BTCUSD", SecurityType.Forex, Market.Bitfinex);
```

Comment:
Ideally we would trust the symbol from incoming requests, such as the order objects or the subscription requests